streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java

```diff
@@ -0,0 +1,376 @@
+        final String taskName = context.taskId().toString();
+        expiredRecordSensor = metrics.storeLevelSensor(
+            taskName,
+            name(),
+            EXPIRED_WINDOW_RECORD_DROP,
+            Sensor.RecordingLevel.INFO
+        );
+        addInvocationRateAndCount(
+            expiredRecordSensor,
+            "stream-" + metricScope + "-metrics",
+            metrics.tagMap("task-id", taskName, metricScope + "-id", name()),
+            EXPIRED_WINDOW_RECORD_DROP
+        );
+
+        if (root != null) {
+            context.register(root, (key, value) -> {
+                put(SessionKeySchema.from(Bytes.wrap(key)), value);
+            });
+        }
+        this.open = true;
+    }
+
+    @Override
+    public void put(final Windowed<Bytes> sessionKey, final byte[] aggregate) {
+        removeExpiredSegments();
+
+        final long windowEndTimestamp = sessionKey.window().end();
+        observedStreamTime = Math.max(observedStreamTime, windowEndTimestamp);
+
+        if (windowEndTimestamp <= observedStreamTime - retentionPeriod) {
+            expiredRecordSensor.record();
+            LOG.debug("Skipping record for expired segment.");
+        } else {
+            if (aggregate != null) {
+                endTimeMap.computeIfAbsent(windowEndTimestamp, t -> new ConcurrentSkipListMap<>());
+                final ConcurrentNavigableMap<Bytes, ConcurrentNavigableMap<Long, byte[]>> keyMap = endTimeMap.get(windowEndTimestamp);
+                keyMap.computeIfAbsent(sessionKey.key(), t -> new ConcurrentSkipListMap<>());
+                keyMap.get(sessionKey.key()).put(sessionKey.window().start(), aggregate);
+            } else {
+                remove(sessionKey);
+            }
+        }
+    }
+
+    @Override
+    public void remove(final Windowed<Bytes> sessionKey) {
+        final ConcurrentNavigableMap<Bytes, ConcurrentNavigableMap<Long, byte[]>> keyMap = endTimeMap.get(sessionKey.window().end());
+        final ConcurrentNavigableMap<Long, byte[]> startTimeMap = keyMap.get(sessionKey.key());
+        startTimeMap.remove(sessionKey.window().start());
+
+        if (startTimeMap.isEmpty()) {
+            keyMap.remove(sessionKey.key());
+            if (keyMap.isEmpty()) {
+                endTimeMap.remove(sessionKey.window().end());
+            }
+        }
+    }
+
+    @Override
+    public byte[] fetchSession(final Bytes key, final long startTime, final long endTime) {
+        removeExpiredSegments();
+
+        if (endTime <= observedStreamTime - retentionPeriod) {
+            return null;
+        }
+
+        final ConcurrentNavigableMap<Bytes, ConcurrentNavigableMap<Long, byte[]>> keyMap = endTimeMap.get(endTime);
+        if (keyMap != null) {
+            final ConcurrentNavigableMap<Long, byte[]> startTimeMap = keyMap.get(key);
+            if (startTimeMap != null) {
+                return startTimeMap.get(startTime);
+            }
+        }
+        return null;
```

Comment:
This has been a point of debate in the past, and not everyone agrees with me, but I'll bring it up just once because it's my honest feedback...

This style of having one-sided conditionals and early returns generally makes it harder to trace the flow of execution. It's more verbose, but also more resistant to bugs to prefer keeping both `if` and `else` branches (or using ternary operators), and (again, generally) not returning early or falling though to a catch-all return at the end of a method.

This feedback applies elsewhere in this PR as well.

Reply:
Hm, I see your point. I think this is unavoidable in places but some other methods (hasNext) can be cleaned up